core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java

```diff
@@ -971,4 +980,147 @@ public void close() throws IOException {
     public ThreadContext getThreadContext() {
         return threadContext;
     }
+
+    static class ReschedulingRunnable implements Runnable {
```

Comment:
can we use AbstractRunnable ? this will give us handy onAfter and onFailure methods.